activerecord/lib/active_record/connection_adapters/postgresql/schema_definitions.rb

```
@@ -261,6 +261,17 @@ def unique_constraint(column_name, **options)
           unique_constraints << new_unique_constraint_definition(column_name, options)
         end
 
+        def check_constraint(expression, **options)
+          if options[:validate] == false
+            ActiveRecord.deprecator.warn(<<~MSG)
+              Providing `validate: false` for `check_constraint` is deprecated because it is ignored by the PostgreSQL.
+              Use the `add_check_constraint` with `validate: false` to create a check constraint that is `NOT VALID`.
```

The `NOT VALID` option is not only for CHECK constraints, so why is it being deprecated only for CHECK constraints?

Also, the `NOT VALID` option is currently just ignored, how will it behave in the future after going through a deprecation cycle? Will it cause an error?

---

Since #53732 was about the check constraint, I focused on that and forgot about the foreign key. I have now applied the warnings for both foreign keys and check constraints. Additionally, I have updated the deprecation warnings to indicate that this option will be removed in Rails 8.2.

---

> I have updated the deprecation warnings to indicate that this option will be removed in Rails 8.2.

What behavior change does "this option will be removed" mean?
Currently the option effectively does nothing (it behaves like it's being ignored).
Will it be changed to something different?

---

> What behavior change does "this option will be removed" mean?

I have updated the deprecation error message that this option will raise a argument error.

---

> If we frequently make mistakes with `CONSTRAINT` options, it might be useful to validate the options like https://github.com/rails/rails/pull/46178.

As I said at https://github.com/rails/rails/issues/53732#issuecomment-2502868227, how about making all options which are ignored an error instead of making only one option an error?

Although #27756 was introduced in Rails 5.2, no one has reported any issues with it.

The reason is that, as quoted in https://github.com/rails/rails/issues/53732#issuecomment-2502868227, the `NOT VALID` option is not practically useful unless the table has a large number of records or is updated frequently, so there is no need for users to explicitly specify it when creating a table.

#53732 was reported because Rails was dumping schema information in a non-restorable state, and that issue has already been resolved in #53735.

Is it worth taking care of just one option like this that users never explicitly specify?